onnxruntime/contrib_ops/webgpu/quantization/matmul_nbits.cc

```diff
@@ -537,60 +540,111 @@ Status MatMulNBitsWideTileProgram::GenerateShaderCode(ShaderHelper& shader) cons
+    let offset = row * uniforms.input_b_stride[0] + col * uniforms.input_b_stride[1];
+
+    // u32 holds 8 packed uint4.
+    let array_index = offset / 8u;
+    let component_index = offset % 8u;
+    let packed_value = zero_points[array_index];
+
+    // Extract the uint4 component
+    let shift_amount = component_index * 4u;
+    let masked_value = (packed_value >> shift_amount) & 0xFu;
+
+    return output_element_t(masked_value);
+  }
+  return output_element_t(0);
+}
+)";
+  } else {
+    shader.AdditionalImplementation() << R"(
+fn mm_read_zero(row : u32, col : u32) -> output_element_t {
+  // The default zero point is 8.
+  return output_element_t(8);
+}
+)";
+  }
+
+  shader.AdditionalImplementation() << "\n"
+                                    << "fn mm_write_y(batch : u32, row : u32, col : u32, value : output_value_t) {\n"
+                                    << "  if (row < uniforms.output_shape[1] && col < uniforms.output_shape[2]) {\n"
+                                    << "    " << y.SetByIndices("output_indices_t(batch, row, col)", "value") << "\n"
+                                    << "  }\n"
+                                    << "}\n";
+
+  shader.AdditionalImplementation() << R"(
+fn dequantize_packed8xU4(packed_value : u32, zero_point : output_element_t, scale : output_element_t) -> mat2x4<output_element_t> {
+  let lower_values: vec4<u32> = unpack4xU8(packed_value & 0x0F0F0F0Fu);
+  let upper_values: vec4<u32> = unpack4xU8((packed_value >> 4u) & 0x0F0F0F0Fu);
+
+  let zero_matrix: mat2x4<output_element_t> = mat2x4<output_element_t>(
+      zero_point, zero_point, zero_point, zero_point,
+      zero_point, zero_point, zero_point, zero_point
+  );
+
+  var dequantized_values: mat2x4<output_element_t> = mat2x4<output_element_t>(
+      output_element_t(lower_values[0]), output_element_t(upper_values[0]),
+      output_element_t(lower_values[1]), output_element_t(upper_values[1]),
+      output_element_t(lower_values[2]), output_element_t(upper_values[2]),
+      output_element_t(lower_values[3]), output_element_t(upper_values[3])
+  );
+
+  dequantized_values = (dequantized_values - zero_matrix) * scale;
+  return dequantized_values;
+}
+)";
 
   // declare const variables
-  shader.AdditionalImplementation() << "\n";
-  shader.AdditionalImplementation() << "// A block32 containing 8 components of `a`." << "\n";
-  shader.AdditionalImplementation() << "const kAComponentsForBlock32 = " << kAComponentsForBlock32 << "u;\n";
-  shader.AdditionalImplementation() << "const tile_m = " << tile_m_ << "u;\n";
-  shader.AdditionalImplementation() << "const tile_n = " << tile_n_ << "u;\n";
+  shader.AdditionalImplementation() << "\n"
+                                    << "// A block32 containing 8 components of `a`." << "\n"
+                                    << "const kAComponentsForBlock32 = " << kAComponentsForBlock32 << "u;\n"
+                                    << "const kTileM = " << tile_m_ << "u;\n"
```

Comment:
nit: No need to change the naming style. In shader, we use snake_case for variables.